### PR TITLE
HIP port of #103: optimized int8 attention

### PR DESCRIPTION
**Pull request by 0xDELUXA**
## Overview

Correctness/performance port of #103 to the HIP backend, enabling `int8_attention` on AMD RDNA3, 3.5, 4. It declines on RDNA2, which has no matrix cores. The main performance gains are seen on RDNA4.

The kernel is a separate implementation, not a direct translation. RDNA's WMMA layout is transposed relative to `mma.m16n8k16`, so the kernel computes `S^T = K @ Q^T` and builds `O^T = V^T @ P^T`. This keeps the online softmax per-lane and, on gfx12, lets the score accumulator feed directly into the second MMA without going through shared memory. gfx11 needs two shuffles and four `v_perm_b32` instructions instead.

The fragment and operand layout is inspired by "Add native gfx12 attention backend" by @jammm (thu-ml/SageAttention#368).

Both matmuls remain int8, as in #103, so the implementation does not depend on fp8 WMMA and RDNA3 is supported as well as RDNA4.

Quantization keeps the CUDA numerics while adapting scale granularity to the RDNA fragment layout: one scale per query row and one per 16 adjacent keys. V needs no key permutation.

## Performance

Several additional optimizations were tested and reverted because they regressed performance or caused spills. A smaller performance change can be expected on RDNA3, while RDNA2 is unsupported because it has no matrix cores.

Measured on gfx1200 against `torch.nn.functional.scaled_dot_product_attention` with `TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1`, best of seven alternating rounds:

- 1.35x at 4K keys, `head_dim=128`
- 1.55x at 8K
- 1.66x on 4K causal
- 2.10x at `head_dim=256`
- 1.14x on Wan-shaped 14040x512 cross attention

The attention kernel alone is 1.67-1.79x faster; quantization accounts for the remaining 15-20%. Below ~2K sequence length it is slower than SDPA due to host overhead and multiple kernel launches/allocations.

## Validation

`tests/test_int8_attention.py` passes unchanged apart from the compute-capability gating test, which now has a HIP counterpart. A separate 71-case sweep covering different head dimensions, dtypes, masks, causal/GQA/ragged inputs, `smooth_k`, `convrot` and split prequantization stays within 0.015 NRMSE of fp32 SDPA, with fused and split paths bitwise identical.

The kernel compile-verifies for all 17 architectures in `architectures.json` as a single fat binary, with no scratch spill on gfx12. I runtime-verified and benchmarked RDNA4, while @tvukovic-amd validated RDNA3/3.5.

## Trying it

The CI wheels include the HIP backend precompiled for every target in `architectures.json`, so no source build is required. Install the artifact for your platform and use it with ComfyUI. It additionally needs the `--use-ck-attention` flag.

### CODE REVIEW

**Comment by tvukovic-amd** [comment]
@0xDELUXA Tested on RDNA3/3.5, here are the results:

gfx1100 - 7900 XTX 24GB:

- `pytest tests/test_int8_attention.py - 5 passed, 54 skipped`
- perf tests
```
shape                                SDPA ms   int8 ms  speedup  jitter   NRMSE
--------------------------------------------------------------------------------
b1 h24 q1024 k1024 d64                 0.112     0.192    0.58x   9.2%  0.0140
b1 h24 q4096 k4096 d64                 1.754     1.782    0.98x   3.1%  0.0146
b1 h16 q1024 k1024 d128                0.174     0.249    0.70x   2.6%  0.0144
b1 h16 q2048 k2048 d128                0.578     0.666    0.87x   0.7%  0.0146
b1 h16 q4096 k4096 d128                2.138     2.130    1.00x   0.5%  0.0148
b1 h16 q8192 k8192 d128                8.269     8.146    1.02x   0.5%  0.0152
b1 h16 q4096 k4096 d128 causal         1.513     1.405    1.08x   0.6%  0.0138
b2 h12 q14040 k512 d128                1.635     1.775    0.92x   0.3%  0.0140
b1 h8 q4096 k4096 d256                 2.830     2.343    1.21x   0.3%  0.0153
b1 h16 q2048 k2048 d128 masked         1.533     1.216    1.26x   1.7%  0.0144
b1 h8 q2048 k2048 d256 masked          1.821     1.785    1.02x   0.8%  0.0149

geomean 0.946x over 11 shapes
Shapes below ~2K tokens are host bound (4 kernel launches and 7 allocations
against SDPA's one), so they jitter hard and are expected to lose.

shape                               quant ms   attn ms  attn vs SDPA
--------------------------------------------------------------------
b1 h16 q1024 k1024 d128                0.083     0.165         1.09x
b1 h16 q2048 k2048 d128                0.112     0.572         1.04x
b1 h16 q4096 k4096 d128                0.165     2.004         1.10x
b1 h16 q8192 k8192 d128                0.495     7.823         1.08x
```

gfx1151 - Strix Halo 128GB:

- `pytest tests/test_int8_attention.py - 5 passed, 54 skipped`
- perf tests
```
shape                                SDPA ms   int8 ms  speedup  jitter   NRMSE
--------------------------------------------------------------------------------
b1 h24 q1024 k1024 d64                 0.290     0.334    0.87x   0.8%  0.0141
b1 h24 q4096 k4096 d64                 4.054     4.039    1.00x   5.4%  0.0144
b1 h16 q1024 k1024 d128                0.357     0.428    0.83x   0.7%  0.0144
b1 h16 q2048 k2048 d128                1.444     1.827    0.79x  45.3%  0.0147
b1 h16 q4096 k4096 d128                8.042     6.609    1.22x  34.9%  0.0150
b1 h16 q8192 k8192 d128              410.501    39.197   10.47x   4.2%  0.0151
b1 h16 q4096 k4096 d128 causal         6.447     4.271    1.51x 107.5%  0.0138
b2 h12 q14040 k512 d128                4.111     9.232    0.45x  11.3%  0.0141
b1 h8 q4096 k4096 d256                 9.504    11.698    0.81x   2.9%  0.0154
b1 h16 q2048 k2048 d128 masked        15.783    10.290    1.53x  15.8%  0.0144
b1 h8 q2048 k2048 d256 masked          7.992     5.211    1.53x   0.4%  0.0148

geomean 1.226x over 11 shapes
Shapes below ~2K tokens are host bound (4 kernel launches and 7 allocations
against SDPA's one), so they jitter hard and are expected to lose.

shape                               quant ms   attn ms  attn vs SDPA
--------------------------------------------------------------------
b1 h16 q1024 k1024 d128                0.120     0.337         1.05x
b1 h16 q2048 k2048 d128                0.438     1.179         1.78x
b1 h16 q4096 k4096 d128                2.479     5.544         1.92x
b1 h16 q8192 k8192 d128                8.307    22.639        18.85x
```

gfx1103 - Hawk Point 16GB:

- `pytest tests/test_int8_attention.py - 5 passed, 54 skipped`
- perf tests
```
shape                                SDPA ms   int8 ms  speedup  jitter   NRMSE
--------------------------------------------------------------------------------
b1 h24 q1024 k1024 d64                 1.676     1.365    1.23x   2.8%  0.0141
b1 h24 q4096 k4096 d64                31.556    18.116    1.74x   2.4%  0.0145
b1 h16 q1024 k1024 d128                1.880     1.900    0.99x   7.7%  0.0143
b1 h16 q2048 k2048 d128                8.019     7.729    1.04x   1.6%  0.0147
b1 h16 q4096 k4096 d128               32.705    26.631    1.23x   4.8%  0.0148
b1 h16 q8192 k8192 d128              118.376    96.257    1.23x  17.1%  0.0151
b1 h16 q4096 k4096 d128 causal        36.992    20.036    1.85x   1.2%  0.0137
b2 h12 q14040 k512 d128               31.635    16.146    1.96x   2.7%  0.0141
b1 h8 q4096 k4096 d256                84.292    22.129    3.81x   2.5%  0.0152
b1 h16 q2048 k2048 d128 masked        12.411     9.771    1.27x   3.2%  0.0144
b1 h8 q2048 k2048 d256 masked         22.302    22.758    0.98x   2.3%  0.0148

geomean 1.444x over 11 shapes
Shapes below ~2K tokens are host bound (4 kernel launches and 7 allocations
against SDPA's one), so they jitter hard and are expected to lose.

shape                               quant ms   attn ms  attn vs SDPA
--------------------------------------------------------------------
b1 h16 q1024 k1024 d128                0.586     1.082         1.55x
b1 h16 q2048 k2048 d128                1.732     3.805         1.35x
b1 h16 q4096 k4096 d128                6.538    14.618         2.35x
b1 h16 q8192 k8192 d128               18.399    64.933         1.74x
```

**Comment by 0xDELUXA** [comment]
Pushed a change: on RDNA3, misaligned shared-memory rows made each K/V fragment read cost 4 LDS instructions instead of 1. Fixing that halves LDS instructions (256 -> 128), cuts total instructions 8.5%, improves head_dim 64 occupancy from 7 to 9 waves/SIMD, and halves the worst register spill. gfx12 is unchanged.

It's compile-verified only for RDNA3. The same pytest suite and benchmark need to be re-run on actual RDNA3 hardware.

### Regarding the tests

5 passed, 54 skipped means nothing actually ran, the checkout shadows the installed wheel, so the extension isn't found and every GPU test skips. A correct run is 54 passed, 5 skipped.

Check first, from the repo root:

```bash
python -c "import comfy_kitchen as ck; print(ck.__file__, ck.int8_attention_is_available())"
```

If it prints `False`, build the extension into that same checkout (use your own arch):

```bash
COMFY_HIP_ARCHS=gfxarch COMFY_KITCHEN_BUILD_HIP=1 python setup.py build_ext --inplace
```

Then run:

```bash
pytest tests/test_int8_attention.py
```

The 5 skips are the CUDA compute-capability test, which doesn't apply on ROCm.

**Comment by astrelsky** [comment]
```
device      AMD Radeon(TM) 890M Graphics  (gfx1150)
torch       2.14.0a0+rocm7.14.0a20260618  hip=7.14.60850 cuda=None
aotriton    TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1
method      best of 7 alternating rounds, 30 iters each, bf16

shape                                SDPA ms   int8 ms  speedup  jitter   NRMSE
--------------------------------------------------------------------------------
b1 h24 q1024 k1024 d64                 0.992     1.243    0.80x  13.5%  0.0141
b1 h24 q4096 k4096 d64                15.923    16.418    0.97x   7.0%  0.0145
b1 h16 q1024 k1024 d128                1.611     1.491    1.08x   1.6%  0.0144
b1 h16 q2048 k2048 d128                4.205     4.888    0.86x   5.4%  0.0146
b1 h16 q4096 k4096 d128               32.119    19.273    1.67x   7.3%  0.0147
b1 h16 q8192 k8192 d128              221.362    79.590    2.78x   2.5%  0.0150
b1 h16 q4096 k4096 d128 causal        29.983    12.334    2.43x   1.1%  0.0136
b2 h12 q14040 k512 d128               24.795    13.506    1.84x   1.1%  0.0141
b1 h8 q4096 k4096 d256                66.581    17.874    3.72x   4.8%  0.0153
b1 h16 q2048 k2048 d128 masked        10.037     9.143    1.10x   2.0%  0.0145
b1 h8 q2048 k2048 d256 masked         24.240     6.545    3.70x   2.3%  0.0148

geomean 1.636x over 11 shapes
Shapes below ~2K tokens are host bound (4 kernel launches and 7 allocations
against SDPA's one), so they jitter hard and are expected to lose.

shape                               quant ms   attn ms  attn vs SDPA
--------------------------------------------------------------------
b1 h16 q1024 k1024 d128                0.513     0.964         1.67x
b1 h16 q2048 k2048 d128                1.389     3.664         1.15x
b1 h16 q4096 k4096 d128                5.391    13.143         2.39x
b1 h16 q8192 k8192 d128               14.566    56.405         3.99x
```

@tvukovic-amd `b1 h16 q8192 k8192 d128              410.501    39.197   10.47x   4.2%  0.0151` sdpa on the Strix Halo might be worth investigating. It looks like it was almost twice as fast on Strix Point, which definitely shouldn't be.

**Comment by LuXuxue** [comment]
This is the report that deepseek port sageattention(base on https://github.com/thu-ml/SageAttention/pull/368) to rdna3. Now it run as fast as triton backend on gfx1103. Hope this report can make int8 attention run faster. The full sageatten code is [here](https://github.com/LuXuxue/sageattention-rdna3).

[NativeBackendOptimizeReport-en.md](https://github.com/user-attachments/files/30923406/NativeBackendOptimizeReport-en.md)

**Comment by 0xDELUXA** [comment]
> This is the report that deepseek port sageattention(base on [thu-ml/SageAttention#368](https://github.com/thu-ml/SageAttention/pull/368)) to rdna3. Now it run as fast as triton backend on gfx1103. Hope this report can make int8 attention run faster. The full sageatten code is [here](https://github.com/LuXuxue/sageattention-rdna3).
> 
> [NativeBackendOptimizeReport-en.md](https://github.com/user-attachments/files/30923406/NativeBackendOptimizeReport-en.md)

Thanks for the pointer. I cloned it and read through it.

The key difference is the PV matmul:

|             | sageattention-rdna3<br>(port of https://github.com/thu-ml/SageAttention/pull/368 to RDNA3) | this PR<br>(port of #103) |
|-------------|---------------------------------------------------------------------------------------------|------------------------|
| QK          | int8 WMMA                                                                                    | int8 WMMA              |
| PV          | fp16 WMMA                                                                                    | int8 WMMA              |
| V storage   | fp16                                                                                         | int8                   |
| V transpose | separate kernel                                                                              | fused into V quantizer |

sageattention-rdna3's PV uses `wmma_f32_f16`, and V stays fp16 with a separate transpose. The 2x int8-vs-fp16 WMMA figure suggests the PV is roughly half the rate, with twice the V bytes and an extra transpose pass.

The one idea I did port was `v_permlanex16_b32` instead of `__shfl_xor`. On gfx11, `__shfl_xor` involves `ds_bpermute` + `s_wait_dscnt`, while `permlanex16` is a single VALU op. It's now gated to gfx11; gfx12 is byte-identical before/after.

What would actually settle this is benchmarking both kernels on the same GPU. If you're willing to run that, I'd like to see the numbers.

One caveat: benchmark both in the same process/session with a few hundred warmup iterations. Comparing separate builds/processes gave me a fake 3-54% swing.

**Comment by 0xDELUXA** [comment]
@coderabbitai resume
